File: LeetCode200--400/206.反转链表.cs
/*
 * @lc app=leetcode.cn id=206 lang=csharp
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
            if (head == null || head.next == null)
            {
                return head;
            }
            ListNode tempNode = ReverseList(head.next);
            head.next.next = head;
            head.next = null;
            return tempNode;
    }
}
// @lc code=end

File: LeetCode200--400/206.反转链表.cpp
/*
 * @lc app=leetcode.cn id=206 lang=cpp
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
//[AC]
    ListNode* GMT1(ListNode* head)
    {
        if(head == nullptr || head->next == nullptr)
            return head;
        ListNode* newhead = GMT1(head->next);
        head->next->next = head;
        head->next = nullptr;
        return newhead;
    }
    //[AC]
    ListNode* GMT2(ListNode* head)
    {
        if(head == nullptr || head->next == nullptr)
            return head;
        ListNode* pre = nullptr;
        ListNode* cur = head;
        ListNode* next = nullptr;
        while(cur)
        {
            next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
    ListNode* reverseList(ListNode* head) {
       // return GMT1(head);
       return GMT2(head);
    }
};
// @lc code=end

